### PR TITLE
#799: fix zip extraction to preserve file attributes

### LINKED ISSUE #799: binaries from zip download lack executable flags
### Expected behavior

As a IDEasy user, I want to install and run terraform via ide command so that I can easily use it.

### Actual behavior

```
$ ide terraform
Trying to download terraform-1.9.8-mac-x64.zip from https://releases.hashicorp.com/terraform/1.9.8/terraform_1.9.8_darwin_amd64.zip
Downloading 100% │███████████████████████████████████████████████████████████████████████████████████████│ 26/26MiB (0:00:00 / 0:00:00) ?MiB/s
Checksum be591e8c59c49d0cfbc7664d24910a4b43840b89d0a4bbca662149bbf0397e91 is correct.
Unpacking 100% │█████████████████████████████████████████████████████████████████████████████████████████│ 26/26MiB (0:00:00 / 0:00:00) ?MiB/s
Step 'ide (terraform)' failed: java.lang.IllegalStateException: Running command '/Users/hohwille/projects/_ide/software/default/terraform/terraform/1.9.8/terraform' with arguments '-install-autocomplete' failed: Cannot run program "/Users/hohwille/projects/_ide/software/default/terraform/terraform/1.9.8/terraform": error=13, Permission denied
 Step 'Install terraform (*)' failed: java.lang.IllegalStateException: Running command '/Users/hohwille/projects/_ide/software/default/terraform/terraform/1.9.8/terraform' with arguments '-install-autocomplete' failed: Cannot run program "/Users/hohwille/projects/_ide/software/default/terraform/terraform/1.9.8/terraform": error=13, Permission denied
2 step(s) failed out of 3 steps.
An unexpected error occurred!
We are sorry for the inconvenience.
Please check the error below, resolve it and try again.
If the error is not on your end (network connectivity, lack of permissions, etc.) please file a bug:
https://github.com/devonfw/IDEasy/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=IllegalStateException%3A+Running+command+%27%2FUsers%2Fhohwille%2Fprojects%2F_ide%2Fsoftware%2Fdefault%2Fterraform%2Fterraform%2F1.9.8%2Fterraform%27+with+arguments+%27-install-autocomplete%27+failed%3A+Cannot+run+program+%22%2FUsers%2Fhohwille%2Fprojects%2F_ide%2Fsoftware%2Fdefault%2Fterraform%2Fterraform%2F1.9.8%2Fterraform%22%3A+error%3D13%2C+Permission+denied
java.lang.IllegalStateException: Running command '/Users/hohwille/projects/_ide/software/default/terraform/terraform/1.9.8/terraform' with arguments '-install-autocomplete' failed: Cannot run program "/Users/hohwille/projects/_ide/software/default/terraform/terraform/1.9.8/terraform": error=13, Permission denied
	at com.devonfw.tools.ide.process.ProcessContextImpl.run(ProcessContextImpl.java:204)
	at com.devonfw.tools.ide.tool.ToolCommandlet.runTool(ToolCommandlet.java:169)
	at com.devonfw.tools.ide.tool.ToolCommandlet.runTool(ToolCommandlet.java:152)
	at com.devonfw.tools.ide.tool.terraform.Terraform.postInstall(Terraform.java:30)
	at com.devonfw.tools.ide.tool.LocalToolCommandlet.postInstall(LocalToolCommandlet.java:128)
	at com.devonfw.tools.ide.tool.LocalToolCommandlet.install(LocalToolCommandlet.java:103)
	at com.devonfw.tools.ide.tool.ToolCommandlet.runTool(ToolCommandlet.java:166)
	at com.devonfw.tools.ide.tool.ToolCommandlet.runTool(ToolCommandlet.java:152)
	at com.devonfw.tools.ide.tool.ToolCommandlet.runTool(ToolCommandlet.java:139)
	at com.devonfw.tools.ide.tool.ToolCommandlet.run(ToolCommandlet.java:130)
	at com.devonfw.tools.ide.context.AbstractIdeContext.applyAndRun(AbstractIdeContext.java:860)
	at com.devonfw.tools.ide.context.AbstractIdeContext.run(AbstractIdeContext.java:767)
	at com.devonfw.tools.ide.cli.Ideasy.runOrThrow(Ideasy.java:89)
	at com.devonfw.tools.ide.cli.Ideasy.run(Ideasy.java:52)
	at com.devonfw.tools.ide.cli.Ideasy.main(Ideasy.java:29)
	at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: java.io.IOException: Cannot run program "/Users/hohwille/projects/_ide/software/default/terraform/terraform/1.9.8/terraform": error=13, Permission denied
	at java.base@21.0.2/java.lang.ProcessBuilder.start(ProcessBuilder.java:1170)
	at java.base@21.0.2/java.lang.ProcessBuilder.start(ProcessBuilder.java:1089)
	at com.devonfw.tools.ide.process.ProcessContextImpl.run(ProcessContextImpl.java:173)
	... 15 more
Caused by: java.io.IOException: error=13, Permission denied
	at java.base@21.0.2/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base@21.0.2/java.lang.ProcessImpl.<init>(ProcessImpl.java:295)
	at java.base@21.0.2/java.lang.ProcessImpl.start(ProcessImpl.java:225)
	at java.base@21.0.2/java.lang.ProcessBuilder.start(ProcessBuilder.java:1126)
	... 17 more
```

### Steps to reproduce (bug) / Use Case of feature request (enhancement)

1. `ide terraform`

### Related/Dependent Issues

Same bug also exists for gradle.
Update: same error also for `ide gh -v`

### Comments/Hints:

If we download tools as ZIP on Linux or Mac, we must call `chmod +x` on the tool binary after unpacking.

### Affected version:

* OS: Mac
* 2024.12.001-beta-SNAPSHOT

### PR DESCRIPTION
**Pull request by hohwille**
fixes #799 

I first messed around with `ZipEntry.getExtra()` see [this thread](https://stackoverflow.com/questions/54654170/what-are-the-extra-bytes-in-the-zipentry-used-for) but had to give up.

Then I started from scratch and found [this promising issue](https://bugs.openjdk.org/browse/JDK-8213082) and [this promising page](https://docs.oracle.com/javase/8/docs/technotes/guides/io/fsp/zipfilesystemprovider.html).

Massive simplification of our code whilst also fixing all preservation of file permissions (not only executable flags but all attributes).

### CODE REVIEW

**Comment by hohwille** [comment]
Tested on Linux:
```
Trying to download terraform-1.10.1-linux-x64.zip from https://releases.hashicorp.com/terraform/1.10.1/terraform_1.10.1_linux_amd64.zip
Downloading 100% │████████████████████████│ 26/26MiB (0:00:00 / 0:00:00) ?MiB/s
```
And
```
hohwille@CE49454:~/projects/IDEasy/software/terraform$ ls -lah ~/Downloads/ide/default/terraform-1.10.1-linux-x64.zip
-rw-r--r-- 1 hohwille hohwille 27M Dec  5 23:32 /home/hohwille/Downloads/ide/default/terraform-1.10.1-linux-x64.zip
```

That looks sane even though 26 vs. 27 but I assume this is MB (1024^2) vs. MiB (1000^2).

But currently I cannot know the extracted size but also I now do not have a clue about uncompressed file sizes during unpacking.
Therefore the progressbar for unpacking is kind of confusing/wrong:
```
Checksum d49d4d08ed092a8dec335f1fe3e127d1a285f160557323dd5a84ddc0c8472e1a is correct.
Unpacking 100% │██████████████████████████│ 34/34MiB (0:00:00 / 0:00:00) ?MiB/s
Successfully installed terraform in version 1.10.1
```
And
```
hohwille@CE49454:~/projects/IDEasy/software/terraform$ ls -lah
total 87M
drwxr-xr-x 2 hohwille hohwille 4.0K Dec  5 23:32 .
drwxr-xr-x 3 hohwille hohwille 4.0K Dec  5 23:32 ..
-rw-r--r-- 1 hohwille hohwille    6 Dec  5 23:32 .ide.software.version
-rw-r--r-- 1 hohwille hohwille 4.9K Dec  5 23:32 LICENSE.txt
-rwxr-xr-x 1 hohwille hohwille  87M Dec  5 23:32 terraform
```

We can test if we can first summarize the sizes but we would need to traverse the virtual JAR/ZIP filesystem twice then.
This could lead to performance issues.
I will do a test with a large archive to measure the effect.
If Java VFS will do an efficient caching, that should be fine and we could then provide an accurate progress bar.

**Comment by hohwille** [comment]
Progressbar entirely refactored and now correct for extract ZIP/JAR. Overhead when unpacking intellij ZIP (largest example we have) was < 1 second (increased from 22 sec to ~23 sec).

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java [hunk 1]
```diff
@@ -614,7 +601,37 @@ private Path getProperInstallationSubDirOf(Path path, Path archiveFile) {
...
-    extractZipArchive(file, targetDir);
+    this.context.info("Extracting ZIP file {} to {}", file, targetDir);
+    URI uri = URI.create("jar:" + file.toUri());
+    try (FileSystem fs = FileSystems.newFileSystem(uri, FS_ENV)) {
+      long size = 0;
+      for (Path root : fs.getRootDirectories()) {
+        size += getFileSizeRecursive(root);
+      }
+      try (IdeProgressBar bp = this.context.newProgressbarForExtracting(size)) {
+        PathCopyListener listener = (source, target, directory) -> {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:612]: Can you please explain this in our next KT-Session?

File: cli/src/main/java/com/devonfw/tools/ide/io/IdeProgressBar.java [hunk 2]
```diff
@@ -5,6 +5,36 @@
...
+   */
+  long getMaxSize();
+
+  /**
+   * @return the name of the unit displayed to the end user (e.g. "files" or "MiB").
+   */
+  String getUnitName();
+
+  /**
+   * @return the size of a single unit (e.g. 1 if the {@link #stepBy(long) reported progress} and {@link #getMaxSize() max size} numbers as is or 1000 for
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/io/IdeProgressBar.java:33]:
```suggestion
   * @return the size of a single unit (e.g. 1 if the {@link #stepBy(long) reported progress} and {@link #getMaxSize() max size} numbers remain unchanged or 1000 for
```

File: cli/src/main/java/com/devonfw/tools/ide/io/PathCopyListener.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @see FileAccess#copy(Path, Path, FileCopyMode, PathCopyListener)
+ */
+@FunctionalInterface
+public interface PathCopyListener {
+
+  /**
+   * An empty {@link PathCopyListener} instance doing nothing.
+   */
+  static PathCopyListener NONE = (s, t, d) -> {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/io/PathCopyListener.java:16]: ??? Please elaborate.

**Review by jan-vcapgemini — APPROVED** [review]
Co-Reviewed by: @WorkingAmeise Looks good to us apart from some questions. Approving for faster merge, but please check comments.

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java (hunk 1 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:612]: Sure, but I already did or tried to do in our last "Sprint Review" meeting but maybe we have been to short on time so lets repeat that tomorrow.

File: cli/src/main/java/com/devonfw/tools/ide/io/PathCopyListener.java (hunk 3 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/io/PathCopyListener.java:16]: I do not get the point here? What are you suggesting?
This is an empty implementation that does nothing and it is used for the existing usages that did not have the ability for such a listener before. The JavaDoc is explaining that this is empty and the implementation is obviously empty. So what is the problem?

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/io/PathCopyListener.java:16]: Will also be discussed tomorrow.

**Pull request merged**